xds/internal/balancer/ringhash/ringhash.go

```diff
@@ -37,26 +72,25 @@ type subConn struct {
 	// This is the actual state of this SubConn (as updated by the ClientConn).
 	// The effective state can be different, see comment of attemptedToConnect.
 	state connectivity.State
-	// attemptedToConnect is whether this SubConn has attempted to connect ever.
-	// So that only the initial Idle is Idle, after any attempt to connect,
-	// following Idles are all TransientFailure.
+	// failing is whether this SubConn is in a failing state. A subConn is
+	// considered to be in a failing state if it was previously in
+	// TransientFailure.
 	//
-	// This affects the effective connectivity state of this SubConn, e.g. if
-	// the actual state is Idle, but this SubConn has attempted to connect, the
-	// effective state is TransientFailure.
+	// This affects the effective connectivity state of this SubConn, e.g.
+	// - if the actual state is Idle or Connecting, but this SubConn is failing,
+	// the effective state is TransientFailure.
 	//
-	// This is used in pick(). E.g. if a subConn is Idle, but has
-	// attemptedToConnect as true, pick() will
+	// This is used in pick(). E.g. if a subConn is Idle, but has failing as
+	// true, pick() will
 	// - consider this SubConn as TransientFailure, and check the state of the
 	// next SubConn.
 	// - trigger Connect() (note that normally a SubConn in real
 	// TransientFailure cannot Connect())
 	//
-	// Note this should only be set when updating the state (from Idle to
-	// anything else), not when Connect() is called, because there's a small
-	// window after the first Connect(), before the state switches to something
-	// else.
-	attemptedToConnect bool
+	// A subConn starts in non-failing (failing is false). A transition to
+	// TransientFailure set failing to true (and it stay true). A transition to
+	// Ready sets failing to false.
+	failing bool
```

Comment:
Do we need this bool though?  Can't we encapsulate it all in `state`, and make that the effective state instead of the true state?  I.e. when changing state:

```go
oldState := sc.state
if sc.state != connectivity.TransientFailure || newState == connectivity.Ready || newState == connectivity.Shutdown {
	// If current state is TRANSIENT_FAILURE, do not change except to READY or SHUTDOWN.
	sc.state = newState
}
csEvltr.recordTransition(oldState, sc.state)
```

Reply:
We need to keep the actual state for later (e.g. Connect() needs to know the actual state id Idle)